### PR TITLE
[WIP] fsc.exe - Parallel Parsing

### PR DESCRIPTION
**Pull request by TIHan**
This enables parallel parsing when running `fsc.exe`. I think this should be fairly straight forward to make safe.

### CODE REVIEW

File: src/fsharp/fsc.fs [hunk 1]
```diff
@@ -1841,14 +1842,25 @@ let main0(ctok, argv, legacyReferenceResolver, bannerAlreadyPrinted,
...
+
+            if parallelOptions.MaxDegreeOfParallelism > sourceFiles.Length then
+                parallelOptions.MaxDegreeOfParallelism <- sourceFiles.Length
+
+            let results = Array.zeroCreate sourceFiles.Length
+            Parallel.For(0, sourceFiles.Length, parallelOptions, fun i ->
+                results.[i] <-
+                    let (filename: string, isLastCompiland) = sourceFiles.[i]
+                    let pathOfMetaCommandSource = Path.GetDirectoryName filename
+                    match ParseOneInputFile(tcConfig, lexResourceManager, ["COMPILED"], filename, (isLastCompiland, isExe), errorLogger, (*retryLocked*)false) with
```
Comment (TIHan) [inline src/fsharp/fsc.fs:1858]: The `errorLogger` is probably not concurrent safe; will need to solve that.

**Comment by cartermp** [comment]
How does this stack up in benchmarks?

File: src/fsharp/fsc.fs [hunk 2]
```diff
@@ -1841,14 +1842,25 @@ let main0(ctok, argv, legacyReferenceResolver, bannerAlreadyPrinted,
...
-                | Some input -> Some (input, pathOfMetaCommandSource)
-                | None -> None
-                ) 
+            let sourceFiles = isLastCompiland |> List.zip sourceFiles |> Array.ofSeq
+            
+            let parallelOptions = ParallelOptions()
+            parallelOptions.MaxDegreeOfParallelism <- Environment.ProcessorCount
+
+            if parallelOptions.MaxDegreeOfParallelism > sourceFiles.Length then
+                parallelOptions.MaxDegreeOfParallelism <- sourceFiles.Length
```
Comment (saul) [inline src/fsharp/fsc.fs:1851]:
```suggestion
            let parallelOptions = ParallelOptions(MaxDegreeOfParallelism=min Environment.ProcessorCount sourceFiles.Length)
```

Comment (TIHan) [reply src/fsharp/fsc.fs:1851]: Much better :)

**Comment by TIHan** [comment]
I will get some benchmarks soon. My gut feel is that it will only shave a few seconds off of compiling the compiler as an example.

**Comment by En3Tho** [comment]
This optimization will benefit fcs the most i believe, not fsc. IDE's like VS, Rider already can build stuff really quickly with heuristics and so on, but while working with code fcs is a thing that takes too much time. Can this optimization be turned on for fcs?

**Comment by auduchinok** [comment]
@En3Tho Parsing is usually not a problem in IDEs. To make IDE work faster we need optimizations for type-checking, which is a subsequent phase after the parsing one. Moreover, FCS keeps only few file parsing results, so there's no need to parse files until they're needed by type checker or some IDE feature.

Previously parsing would wait in the FCS queue if there's a long type checking process, and it has been improved in https://github.com/dotnet/fsharp/pull/3601.

**Comment by En3Tho** [comment]
@auduchinok @TIHan Thank you for an answer. Any plans to further parallelize compiler stages? It feels quite sad to wait for fcs every time while having 8/16 core/threads processor and 64 gigs of mem

**Comment by cartermp** [comment]
No current plans, but @tihan has looked into it before. It should be possible in theory without having to rewrite everything, but it won't be easy

**Comment by TIHan** [comment]
@auduchinok is right. Type-checking is where almost all the time is spent. Parsing is really fast in F# - I'm not even sure if parallel parsing files will give us that much benefit (haven't made benchmarks yet).

@En3Tho, @dsyme and I discussed parallel type-checking in FCS last week. We went over a variety of things, and it should be possible.

When I did the compiler server prototype, I made a lot of fixes that supported the work being done in parallel. Perhaps I can re-use some of that work and make small PRs that will inch our way towards parallelism.

Another idea that's interesting is parallel type-checking when compiling: https://github.com/dotnet/fsharp/issues/10214

**Comment by cartermp** [comment]
> I'm not even sure if parallel parsing files will give us that much benefit (haven't made benchmarks yet).

I imagine it would help parsing our FSharp.Core test suite, since other threads could parse stuff while one is working on this big boi: https://github.com/dotnet/fsharp/blob/main/tests/FSharp.Core.UnitTests/FSharp.Core/ComparersRegression.fs

**Comment by KevinRansom** [comment]
@TIHan , it looks like you are not convinced about the performance benefits of this change: https://github.com/dotnet/fsharp/pull/10208#issuecomment-707890156.  So I will close it, please reactivate the issue if you change your mind and get an opportunity to develope it further.

Thanks

**Pull request closed without merging**